odoo/addons/base/models/ir_model.py

```diff
@@ -1733,15 +1733,14 @@ def xmlid_lookup(self, xmlid):
         """Low level xmlid lookup
         Return (id, res_model, res_id) or raise ValueError if not found
         """
-        module, name = xmlid.split('.', 1)
-        xid = self.sudo().search([('module', '=', module), ('name', '=', name)])
-        if not xid:
-            raise ValueError('External ID not found in the system: %s' % xmlid)
-        # the sql constraints ensure us we have only one result
-        res = xid.read(['model', 'res_id'])[0]
-        if not res['res_id']:
+        cr = self.env.cr
+        query = "SELECT id, model, res_id FROM ir_model_data WHERE module=%s AND name=%s"
+        params = module, name = xmlid.split('.', 1)
+        cr.execute(query, params)
+        result = cr.fetchone()
+        if result is None or result[2] is None:
```

Comment:
> Could probably just test on `result[2]` to match the old behavour (where 0 would be the same as missing). Maybe just `if not (result and result[2])`, or `if not result` assuming res_id is always set.

Given that `res_id` is not required, I simply cannot assume that it is always set.

> Leading to: is there a reason why `res_id` is nullable, other than historical? We don't seem to have any record without a res_id properly set in a standard install, and apparently even our DB has no such record (not as part of this but as a separate change, probably with a migration to drop records with no res_id just in case).

As you said, probably historical reasons.